### PR TITLE
[FLINK-10172][table]re-adding asc & desc suffix expression to expressionParser

### PR DESCRIPTION
**Pull request by walterddr**
## What is the purpose of the change

Fix Java expressionParser not correctly resolving `asc` and `desc` keyword.

## Brief change log

  - *Adding back `suffixAsc` and `suffixDesc`*

## Verifying this change

  - *Added table string expression test*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (n/a)

### CODE REVIEW

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/batch/table/stringexpr/CalcStringExpressionTest.scala [hunk 1]
```diff
@@ -85,6 +85,39 @@ class CalcStringExpressionTest extends TableTestBase {
     verifyTableEquals(t1, t2)
   }
 
+  @Test
+  def testSelectWithOrdering(): Unit = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/batch/table/stringexpr/CalcStringExpressionTest.scala]: Can you move these tests into a new class `SortStringExpressionTest`?

**Comment by fhueske** [comment]
merging

**Comment by twalthr** [comment]
@walterddr This PR fixes the issue but the question is why did the parser not resolve the `asc`/`desc` correctly?

**Comment by fhueske** [comment]
`asc` and `desc` are not regular functions. They are indicators for the sort order and converted into special `Ordering` expressions. Without special treatment, they are converted into `Call("asc", fieldRef)` and therefore not correctly handled by `Table.orderBy()` which expects `Ordering` expressions.

This PR adds code that was (accidentially?) dropped when refactoring the `ExpressionParser` in 9244106b334ef54ba3e39a3f2c0c76f46ae4ecd3.

**Comment by walterddr** [comment]
@twalthr yes @fhueske is correct the 2 functions are special expressions which requires special expression parsing. So far I am looking over the diff in https://github.com/apache/flink/pull/1926/files and didn't find anything abnormal but I will add additional tests for the lists of functions originally in the `ExpressionParser` class. Should I open a JIRA for it (haven't find a bug yet)

**Comment by twalthr** [comment]
Thanks for the explanation @fhueske and @walterddr. Feel free to add more tests if they make sense. But actually most of the operators should already have tests.

**Comment by walterddr** [comment]
@twalthr I just did a cross-check and all touched expressions are covered in testings. :-)

**Comment by twalthr** [comment]
Thank you @walterddr. I would open an issue for this minor thing to at least track it.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/ExpressionParser.scala [hunk 2]
```diff
@@ -324,6 +332,8 @@ object ExpressionParser extends JavaTokenParsers with PackratParsers {
     suffixToDate |
     // expression for log
     suffixLog |
+    // expression for ordering
+    suffixAsc | suffixDesc |
```
Comment (yanghua) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/ExpressionParser.scala]: @walterddr when you add the keyword `ASC` and allow it as a suffix, my ASCII scalar function test case : 

```
testAllApis(
      'f0.ascii(),
      "f0.ascii()",                  //here
      "ASCII(f0)",
      "84")
```

will report this error:

```
org.apache.flink.table.api.ExpressionParserException: Could not parse expression at column 7: string matching regex `(?i)\Qas\E' expected but `i' found F0.ascii()
```

After commenting `suffixAsc |` , my test case will pass, but your test will throw an exception.

I think I should rename `ASCII` to other name, such as `ASICII`?  cc @fhueske @twalthr @xccui

Comment (fhueske) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/ExpressionParser.scala]: Naming the function `asicii` does not sound right to me. I'd never look for this name if I'd was looking for ASCII conversion. There should be a proper fix for this. 
Please move this discussion to the ascii function PR.

**Pull request closed without merging**